meteor/sofie-server/data-access/repositories/cache/cached-rundown-repository.ts

```diff
@@ -0,0 +1,48 @@
+import { RundownRepository } from '../interfaces/rundown-repository'
+import { Rundown } from '../../../model/entities/rundown'
+import { MisconfigurationException } from '../../../model/exceptions/misconfiguration-exception'
+import { Identifier } from '../../../model/interfaces/identifier'
+
+export class CachedRundownRepository implements RundownRepository {
+	private static instance: RundownRepository
+
+	public static getInstance(rundownRepository?: RundownRepository): RundownRepository {
+		if (!this.instance) {
+			if (!rundownRepository) {
+				throw new MisconfigurationException(
+					`No RundownRepository provided. Unable to create instance of ${CachedRundownRepository.name}`
+				)
+			}
+			this.instance = new CachedRundownRepository(rundownRepository)
+		}
+		return this.instance
+	}
+
+	private rundownRepository: RundownRepository
+
+	private cachedRundowns: Map<string, Rundown> = new Map()
+
+	constructor(rundownRepository: RundownRepository) {
+		this.rundownRepository = rundownRepository
+	}
+
+	public async getRundown(rundownId: string): Promise<Rundown> {
+		if (!this.cachedRundowns.has(rundownId)) {
+			console.log(`### Rundown with id: "${rundownId}" not found in cache. Loading rundown from database...`)
+			const rundown: Rundown = await this.rundownRepository.getRundown(rundownId)
+			this.cachedRundowns.set(rundownId, rundown)
+		}
+		return this.cachedRundowns.get(rundownId) as Rundown
+	}
+
+	public async getRundownIdentifiers(): Promise<Identifier[]> {
+		return this.rundownRepository.getRundownIdentifiers()
+	}
+
+	// Only save in memory for now
+	public saveRundown(rundown: Rundown): void {
+		if (this.cachedRundowns.has(rundown.id)) {
+			this.cachedRundowns.set(rundown.id, rundown)
```

Comment:
Right now it only saves in memory because I didn't want to write changes to the mongoDb in the demonstration.

_This_ method specifically will not be calling `saveSegment()` etc. It will only call `this.rundownRepository.save()` and then the implementation of that method _might_ call `saveSegment()` etc.

So the implementation of _this_ method would be something like this:
```
public saveRundown(rundown: Rundown): void {
  if (this.cachedRundowns.has(rundown.id)) {
    this.cachedRundowns.set(rundown.id, rundown)
  }
  this.rundownRepository.saveRundown(rundown)
}
```

Whether `RundownRepository.saveRundown()` should be calling `SegmentRepository.saveSegment()` is up to the implementation. There are many ways of implementing the save functionality and that could be a viable one. Personally I would like us to look into the possiblity of using an Object Relationel Manager (ORM) as that can save us a lot of manual work.

Reply:
@andr9528, interesting! It promises [cool stuff](https://mikro-orm.io/docs/unit-of-work#implicit-transactions), but I'm of course a little worried about the performance